core/src/main/java/org/elasticsearch/index/shard/IndexShard.java

```diff
@@ -1571,4 +1571,20 @@ public void onAfter() {
         return false;
     }
 
+    /**
+     * Simple struct encapsulating a shard failure
+     * @see IndexShard#addShardFailureCallback(Callback)
+     */
+    public static final class ShardFailure {
+        public final ShardRouting routing;
+        public final String reason;
+        public final Throwable cause;
```

Comment:
I don't think we should have requirements like this when writing new code... I use getFoo everywhere when I add it even if we are not consistent at least I am not adding any new unneeded annotations..